documentation/development-docs/design-guidelines/piping-best-practices.md

```diff
@@ -241,6 +106,142 @@ public override void ExecuteCmdlet()
+
+```
+Remove-AzSampleResource -InputObject <PSSampleResource> [...] [-PassThru] [-WhatIf] [-Confirm]
+```
+
+For all child resources, the same functionality should be added for the cmdlets above, but for `Get-*` and `New-*` cmdlets, rather than using the `-InputObject` parameter, a separate object parameter should be added to allow piping the object representation of the parent resource.  The implementation of this will be a new parameter set, like the following:
+
+```
+Get-AzChildResource -TopLevelResourceObject <PSTopLevelResource> [...]
+```
+
+#### Long explanation
+This scenario should be used when piping objects around within the same module. For example, if you have a set of `Get-AzFoo`, `Remove-AzFoo`, and `Set-AzFoo` cmdlets, the `Remove-AzFoo` and `Set-AzFoo` cmdlets should have a parameter set that takes the `-InputObject` parameter of type `PSFoo` so that a user can do the following:
+
+```powershell
+# --- Piping scenario ---
+# Setting and removing an individual object
+Get-AzFoo -Name "FooName" -ResourceGroupName "RG" | Set-AzFoo @additionalParams
+Get-AzFoo -Name "FooName" -ResourceGroupName "RG" | Remove-AzFoo
+
+# Setting and removing a collection of objects
+Get-AzFoo | Set-AzFoo @additionalParams
+Get-AzFoo | Remove-AzFoo
+
+
+# --- Non-piping scenario ---
+# Setting and removing an individual object
+$foo = Get-AzFoo -Name "FooName" -ResourceGroupName "RG"
+Set-AzFoo -InputObject $foo @additionalParams
+Remove-AzFoo -InputObject $foo
+
+# Setting and removing a collection of objects
+Get-AzFoo | ForEach-Object { Set-AzFoo -InputObject $_ @additionalParams }
+Get-AzFoo | ForEach-Object { Remove-AzFoo -InputObject $_ }
+```
+
+Another time that this scenario applies is when you have cmdlets for child resources that need information about the parent (top-level) resource. For example you can pipe in the whole parent object to the `New-AzFooBar` and `Get-AzFooBar` cmdlets to get the child resources, and then pipe the child resource object to the `Remove-AzFooBar` and `Set-AzFooBar` cmdlets.
+
+```powershell
+# --- Piping scenario ---
+# Getting all of child resources from all of the parent resources and removing them
+Get-AzFoo | Get-AzFooBar | Remove-AzFooBar
+
+# Getting all of the child resources from all of the parent resources in a resource group and removing them
+Get-AzFoo -ResourceGroupName "RG" | Get-AzFooBar | Remove-AzFooBar
+
+# Getting all of the child resources from a specific parent resource and removing them
+Get-AzFoo -ResourceGroupName "RG" -Name "FooName" | Get-AzFooBar | Remove-AzFooBar
+
+
+# --- Non-piping scenario ---
+# Getting all of the child resources from a specific parent resource and removing them
+$foo = Get-AzFoo -ResourceGroupName "RG" -Name "FooName"
+$fooBar = Get-AzFooBar -InputObject $foo
+Remove-AzFooBar -InputObject $fooBar
+```
```

Comment:
For the exmaple:
```
$fooBar = Get-AzFooBar -InputObject $foo
```
Does it make sense to rename the `-InputObject` to `FooInputObject ` or `FooObject` to be consistent with [this](https://github.com/Azure/azure-powershell/pull/8755/files#diff-c68e2135c55b4af00b8ba10cf532f344R131) line:

```
Get-AzChildResource -TopLevelResourceObject <PSTopLevelResource> [...]
```

Reply:
Good catch -- I updated the examples found in this file.